File: CK3toEU4/Source/CK3World/Religions/Faiths.cpp
#include "Faiths.h"
#include "CommonRegexes.h"
#include "Faith.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "Religions.h"

CK3::Faiths::Faiths(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void CK3::Faiths::registerKeys()
{
	registerRegex(R"(\d+)", [this](const std::string& faithID, std::istream& theStream) {
		auto newFaith = std::make_shared<Faith>(theStream, std::stoll(faithID));
		faiths.insert(std::pair(newFaith->getID(), newFaith));
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

void CK3::Faiths::linkReligions(const Religions& religions, const Titles& titles)
{
	auto counter = 0;
	const auto& religionData = religions.getReligions();
	std::map<std::string, std::string> religiousHeadList; // ID, Title
	for (const auto& title: titles.getTitles())
		if (title.second)
			religiousHeadList.emplace(std::to_string(title.second->getID()), title.first);
	for (const auto& faith: faiths)
	{
		const auto& religionDataItr = religionData.find(faith.second->getReligion().first);
		if (religionDataItr != religionData.end())
		{
			faith.second->loadReligion(*religionDataItr);
			if (religiousHeadList.contains(faith.second->getReligiousHead()))
				faith.second->setReligiousHead(religiousHeadList.at(faith.second->getReligiousHead()));
			++counter;
		}
		else
		{
			throw std::runtime_error(
				 "Faith " + faith.second->getName() + " has religion " + std::to_string(faith.second->getReligion().first) + " which has no definition!");
		}
	}
	Log(LogLevel::Info) << "<> " << counter << " faiths updated.";
}

File: CK3ToEU4/Source/CK3/Religions/Faiths.cs
using System;
using System.Collections.Generic;
using commonItems;
using commonItems.Colors;

namespace CK3ToEU4.CK3.Religions;

class Faiths
{
	public Faiths() {}
	public Faiths(BufferedReader reader, ColorFactory colorFactory)
	{
		var parser = new Parser();
		registerKeys(parser, colorFactory);
		parser.ParseStream(reader);
	}

	public IReadOnlyDictionary<long, Faith> getFaiths() { return faiths; }

	public void linkReligions(Religions religions, Titles.Titles titles)
	{
		int counter = 0;
		var religionData = religions.getReligions();
		Dictionary<string, string> religiousHeadList; // ID, Title
		foreach (var title in titles.getTitles())
		{
			if (title.second)
			{
				religiousHeadList.emplace(std::to_string(title.second->getID()), title.first);
			}
		}
		
		foreach (var faith in faiths)
		{
			const auto& religionDataItr = religionData.find(faith.second->getReligion().first);
			if (religionDataItr != religionData.end())
			{
				faith.second->loadReligion(*religionDataItr);
				if (religiousHeadList.contains(faith.second->getReligiousHead()))
					faith.second->setReligiousHead(religiousHeadList.at(faith.second->getReligiousHead()));
				++counter;
			}
			else
			{
				throw new Exception(
					"Faith " + faith.Value.getName() + " has religion " + faith.Value.getReligion().Key + " which has no definition!");
			}
		}
		Logger.Info("<> " + counter + " faiths updated.");
	}

	private void registerKeys(Parser parser, ColorFactory colorFactory)
	{
		parser.RegisterRegex(CommonRegexes.Integer, (reader, faithId) => {
			var newFaith = new Faith(reader, long.Parse(faithId), colorFactory);
			faiths.Add(newFaith.ID, newFaith);
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}

	private Dictionary<long, Faith> faiths;
}
